src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,646 @@
+  }
+
+  // find all HH\Lib\* namespace use declarations
+  // returns a tuple containing the declaration nodes, the HSL namespace suffixes used
+  protected function findUseDeclarations(
+    vec<INamespaceUseDeclaration> $declarations,
+  ): (vec<INamespaceUseDeclaration>, vec<string>) {
+
+    $search = vec["HH", "Lib"];
+    $nodes = vec[];
+    $suffixes = vec[];
+    foreach ($declarations as $decl) {
+
+      // we only care about "use namespace" directives
+      if (!($decl->getKind() instanceof NamespaceToken)) {
+        continue;
+      }
+
+      if ($decl instanceof NamespaceGroupUseDeclaration) {
+
+        // group declarations: does prefix match?
+        $parts =
+          $decl->getPrefix()->getParts()->getItemsOfType(NameToken::class);
+        if (C\count($parts) !== 2) {
+          continue;
+        }
+        $found_prefix = true;
+        foreach ($parts as $i => $token) {
+          if ($token->getText() === $search[$i]) {
+            continue;
+          }
+
+          $found_prefix = false;
+          break;
+        }
+
+        // prefix didn't match? skip this declaration
+        if (!$found_prefix) {
+          continue;
+        }
+
+        // prefix matches: add node to list of nodes
+        $nodes[] = $decl;
+
+        $clauses = vec(
+          $decl->getClauses()->getDescendantsOfType(NamespaceUseClause::class),
+        );
+        foreach ($clauses as $clause) {
+          $name = $clause->getName();
+          if ($name instanceof NameToken) {
+            $suffixes[] = $name->getText();
+          }
+        }
+      } else {
+
+        invariant(
+          $decl instanceof NamespaceUseDeclaration,
+          'Unhandled declaration type',
+        );
+
+        $clauses =
+          $decl->getClauses()->getItemsOfType(NamespaceUseClause::class);
+        foreach ($clauses as $clause) {
+
+          $name = $clause->getName();
+          if ($name instanceof QualifiedName) {
+            $parts = $name->getParts()->getItemsOfType(NameToken::class);
+            if (C\count($parts) !== 3) {
+              continue;
+            }
+
+            foreach ($parts as $i => $token) {
+              if ($i < 2 && $token->getText() !== $search[$i]) {
+                break;
+              }
+
+              if ($i === 2) {
+                // we found an HH\Lib\* use statement, add the node and suffix
+                $nodes[] = $decl;
+                $suffixes[] = $token->getText();
+              }
+            }
+          }
+        }
+      }
+    }
+    return tuple($nodes, $suffixes);
+  }
+
+  // get a node for a use namespace declaration
+  protected function buildUseDeclaration(
+    vec<string> $suffixes,
+  ): INamespaceUseDeclaration {
+    if (C\count($suffixes) > 1) {
+      // make a grouped use declaration
+      $ns = "{".Str\join($suffixes, ', ')."}";
+    } else {
+      // single namespace use declaration
+      $ns = C\firstx($suffixes);
+    }
+    return $this->nodeFromCode(
```

Comment:
https://github.com/hhvm/hhast/issues/6 would be nice, but could be future work to clean this up, rather than a blocker.

Reply:
That would be awesome (and as I just learned, it'd be good to be able to do that with insertBefore and insertAfter too). I'd feel much better about using a code string if it were part of the library :-).

Let's address separately though since this PR is getting large.